File: arrays/go/solution06.go
package leetcode

func intersect(nums1 []int, nums2 []int) []int {
	countMap := make(map[int]int)
	result := []int{}

	for _, num := range nums1 {
		countMap[num]++
	}

	for _, num := range nums2 {
		if countMap[num] > 0 {
			result = append(result, num)
			countMap[num]--
		}
	}
	return result
}

File: arrays/java/Solution06.java
package arrays.java;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Solution06 {

    public int[] intersect(int[] nums1, int[] nums2) {

        Map<Integer, Integer> countMap = new HashMap<>();
        List<Integer> result = new ArrayList<>();

        for (int num : nums1) {
            countMap.put(num, countMap.getOrDefault(num, 0) + 1);
        }

        for (int num : nums2) {
            if (countMap.containsKey(num) && countMap.get(num) > 0) {
                result.add(num);
                countMap.put(num, countMap.get(num) - 1);
            }
        }

        int[] resArr = new int[result.size()];

        for (int i = 0; i < result.size(); i++) {
            resArr[i] = result.get(i);
        }

        return resArr;
    }
}
